deps/oculus/src/ovrsession.cpp

```diff
@@ -0,0 +1,404 @@
+    desc.Width = bufferSize.w;
+    desc.Height = bufferSize.h;
+    desc.MipLevels = 1;
+    desc.SampleCount = 1;
+    desc.StaticImage = ovrFalse;
+
+    result = ovr_CreateTextureSwapChainGL(session, &desc, &eyes[eye].ColorTextureChain);
+    int length = 0;
+    ovr_GetTextureSwapChainLength(session, eyes[eye].ColorTextureChain, &length);
+
+    if (!OVR_SUCCESS(result)) {
+      std::cout << "Error creating Oculus GL Swap Chain" << std::endl;
+    } else {
+      for (int i = 0; i < length; ++i) {
+        GLuint textureId;
+        ovr_GetTextureSwapChainBufferGL(session, eyes[eye].ColorTextureChain, i, &textureId);
+        glBindTexture(GL_TEXTURE_2D, textureId);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+      }
+    }
+
+    desc.Format = OVR_FORMAT_D32_FLOAT;
+
+    result = ovr_CreateTextureSwapChainGL(session, &desc, &eyes[eye].DepthTextureChain);
+    ovr_GetTextureSwapChainLength(session, eyes[eye].DepthTextureChain, &length);
+
+    if (!OVR_SUCCESS(result)) {
+    } else {
+      for (int i = 0; i < length; ++i) {
+        GLuint textureId;
+        ovr_GetTextureSwapChainBufferGL(session, eyes[eye].DepthTextureChain, i, &textureId);
+        glBindTexture(GL_TEXTURE_2D, textureId);
+
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+      }
+    }
+
+    glGenFramebuffers(1, fboId);
+
+  }
+}
+
+NAN_METHOD(OVRSession::GetRecommendedRenderTargetSize)
+{
+  if (info.Length() != 0)
+  {
+    Nan::ThrowError("Wrong number of arguments.");
+    return;
+  }
+
+  ovrSession session = *ObjectWrap::Unwrap<OVRSession>(info.Holder())->self_;
+  ovrHmdDesc hmdDesc = ObjectWrap::Unwrap<OVRSession>(info.Holder())->hmdDesc;
+
+  // Oculus: Initialize Swap Chain
+  ovrSizei leftEyeTextureSize = ovr_GetFovTextureSize(session, ovrEye_Left, hmdDesc.DefaultEyeFov[ovrEye_Left], 1);
+  ovrSizei rightEyeTextureSize = ovr_GetFovTextureSize(session, ovrEye_Right, hmdDesc.DefaultEyeFov[ovrEye_Right], 1);
+
+  int width  = leftEyeTextureSize.w;
+  int height = std::max(leftEyeTextureSize.h, rightEyeTextureSize.h);
+
+  Local<Object> result = Nan::New<Object>();
+  {
+    Local<String> width_prop = Nan::New<String>("width").ToLocalChecked();
+    Nan::Set(result, width_prop, Nan::New<Number>(width));
+
+    Local<String> height_prop = Nan::New<String>("height").ToLocalChecked();
+    Nan::Set(result, height_prop, Nan::New<Number>(height));
+  }
+  info.GetReturnValue().Set(result);
+}
+
+NAN_METHOD(OVRSession::RequestGetPoses) {
```

Comment:
One thing to keep in mind is that the OpenVR and magic leap bindings [are multi-threaded](https://github.com/webmixedreality/exokit/blob/30921aaf7197defbd8463e88cf1f8f0ff6ca5861/deps/openvr/src/ivrcompositor.cpp#L222), and I believe this one is not.

This means that for Oculus, this binding will hold up the Node render loop and things like network requests would be frame-locked. That's not a bug but it would probably be a notable experience degradation. Additionally, if we merge parallel user threads (https://github.com/webmixedreality/exokit/pull/760), the single-thread solution cannot work at all since it relies on asynchrony of all threads.

The fix for this would be to make sure the pose wait and submit methods are asynchronous and return a promise, while runnig on another thread. The rest of the code would be the same. We don't need to do this before merging, but we'd need to do this soon either way.